fix: remove old subscriptions and subscribers

matt2e opened the pull request:
fixes #1660
fixes #1685

changes:
- add hooks for when we activate and deactivate deployments (`min_replicas > 0` vs `min_replicas == 0`)
- when activating a deployment:
    - upsert subscriptions (maintain state), insert subscribers
    - skip subscriptions with no subscribers (ie: do not insert into db)
- when deactivating a deployment:
    - delete subscriptions still owned by deployment
    - delete subscribers

Skipping subscriptions without subscribers fixes the case where pubsub manager would get stuck trying to progress subscriptions without subscribers, blocking real subscriptions from progressing.

alecthomas commented:
Off the top of my head, inserting new topics/subscriptions when creating the new deployment, and dropping all unused ones in the remove deployment seems reasonable. I don't know how painful that would be.

deniseli commented on backend/controller/dal/dal.go:
```diff
@@ -501,22 +501,51 @@ func (d *DAL) CreateDeployment(ctx context.Context, language string, moduleSchem
 			continue
 		}
-		err := tx.UpsertSubscription(ctx, sql.UpsertSubscriptionParams{
+		// look for subscribers
+		if _, hasSubscriber := slices.Find(moduleSchema.Decls, func(in schema.Decl) bool {
+			v, ok := in.(*schema.Verb)
+			if !ok {
+				return false
+			}
+			_, subscribes := slices.Find(v.Metadata, func(in schema.Metadata) bool {
+				md, ok := in.(*schema.MetadataSubscriber)
+				if !ok {
+					return false
+				}
+				return md.Name == s.Name
+			})
+			return subscribes
```
really nitty nit pick on readability: If you pull this logic into a helper function `hasSubscriber`, you can do:

```
if hasSubscriber(moduleSchema.Decls) {
  // Ignore subscriptions without ...
  continue
}
```

This (1) shortens the for-loop so it's easier to read/comprehend and (2) communicates "look for subscribers" with a func name instead of a comment, which is slightly better for maintainability

matt2e replied:
Yeh agreed, thanks Denise

deniseli commented on backend/controller/sql/queries.sql:
```diff
@@ -645,6 +645,20 @@ UPDATE SET
   topic_id = excluded.topic_id
 RETURNING id;
 
+-- name: DeleteOldSubscriptions :exec
```
nitty nitty naming nit: should this be `DeleteUnusedSubscriptions` or `DeleteObsoleteSubscriptions` since the selection criteria is whether the subscription is used, not whether it's aged past a certain TTL (time to live) date?

matt2e replied:
That's better naming 👍

matt2e replied:
(Renamed again because its changed whats it's doing)

deniseli commented on backend/controller/dal/dal.go:
```diff
@@ -501,22 +501,51 @@ func (d *DAL) CreateDeployment(ctx context.Context, language string, moduleSchem
+	if err := tx.DeleteOldSubscriptions(ctx, moduleSchema.Name, subscriptionNames); err != nil {
+		return model.DeploymentKey{}, fmt.Errorf("could not delete old subscriptions: %w", translatePGError(err))
+	}
+	if err := tx.DeleteSubscribers(ctx, moduleSchema.Name); err != nil {
```
Re:
> These db updates are done in the db commit which inserts the deployment. At that time the previous deployment is still live and the new deployment has min_replicas set to 0. If the next commit (ReplaceDeployment) does not complete then the old deployment will be in a bad state (all subscribers stop receiving events).

Is it possible to refactor this code to combine the 2 commits into a single [Postgres Transaction](https://www.postgresql.org/docs/current/tutorial-transactions.html)? That basically turns it into a single atomic operation, so that if something goes wrong with the second step, then the first will not be committed either. I don't know this code very well, so that may well be a bad idea :p

matt2e replied:
I think there's history in this area of code for why we create a deployment in one transaction, and then replace the old deployment with the new deployment in a second commit. I think @wesbillman mentioned when I first joined that this area had recently gone through some hardening to catch glitches?

deniseli replied:
Oh interesting...

matt2e commented:
> Off the top of my head, inserting new topics/subscriptions when creating the new deployment, and dropping all unused ones in the remove deployment seems reasonable. I don't know how painful that would be.

ended up doing this 👍 It requires an extra column in `subscriptions` to point to the latest deployment it is part of

matt2e commented:
@alecthomas ok that didn't exactly work. Here's the bad case steps:
- `ftl dev`
- Initial deploy, creates pubsub records
- change code, save
    - `CreateDeployment()` inserts the deployment, insert/update subscriptions/subscribers
    - ReplaceDeployment gets called, deletes old subscriptions/subscribers
- undo code changes, save
    - `CreateDeployment()` exits early because we detect that the original deployment matches what we need
        - we do not insert/update the subscriptions.subscribers
    - ReplaceDeployment gets called, deletes old subscriptions/subscribers
- So we end up with no subscriptions/subscribers for the module

So the solution has evolved a bit:
- Added hooks for `deploymentWillActivate` and `deploymentWillDeactivate` in `dal/dal.go`
    - called by `ReplaceDeployment` and `SetDeploymentReplicas`
- Moved all this pubsub specific code to `dal/pubsub.go`
- This also handles cases like modules being removed and then re-added (eg by changing branches)
